handlers/messagebird/messagebird.go

```diff
@@ -0,0 +1,308 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+	clog.Error(courier.ErrorExternal(fmt.Sprint(receivedStatus.StatusErrorCode), "EC_SUBSCRIBER_OPTEDOUT"))
+
+	return handlers.WriteMsgStatusAndResponse(ctx, h, channel, status, w, r)
+}
+
+func (h *handler) receiveMessage(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request, payload *ReceivedMessage, clog *courier.ChannelLog) ([]courier.Event, error) {
+	err := h.validateSignature(channel, r)
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+
+	// no message? ignore this
+	if payload.Body == "" && !payload.MMS {
+		return nil, handlers.WriteAndLogRequestIgnored(ctx, h, channel, w, r, "Ignoring request, no message")
+	}
+
+	// create our date from the timestamp
+	date := payload.CreatedDatetime
+
+	// create our URN
+	urn, err := handlers.StrictTelForCountry(payload.Originator, channel.Country())
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+	text := payload.Body
+
+	// build our msg
+	msg := h.Backend().NewIncomingMsg(channel, urn, text, payload.ID, clog).WithReceivedOn(date.UTC())
+
+	// process any attached media
+	if payload.MMS {
+		for _, mediaURL := range payload.MediaURLs {
+			msg.WithAttachment(mediaURL)
+		}
+	}
+	// and finally write our message
+	return handlers.WriteMsgsAndResponse(ctx, h, []courier.Msg{msg}, w, r, clog)
+}
+
+func (h *handler) Send(ctx context.Context, msg courier.Msg, clog *courier.ChannelLog) (courier.MsgStatus, error) {
+
+	authToken := msg.Channel().StringConfigForKey(courier.ConfigAuthToken, "")
+	if authToken == "" {
+		return nil, fmt.Errorf("missing config 'auth_token' for Messagebird channel")
+	}
+
+	user := msg.URN().Path()
+	status := h.Backend().NewMsgStatusForID(msg.Channel(), msg.ID(), courier.MsgErrored, clog)
+
+	// create base payload
+	payload := &Message{
+		Recipients: []string{user},
+		Originator: msg.Channel().Address(),
+		Reference:  msg.ID().String(),
+	}
+	// build message payload
+
+	if len(msg.Text()) > 0 {
+		payload.Body = msg.Text()
+	}
+	sendUrl := ""
+	if len(msg.Attachments()) > 0 {
+		sendUrl = mmsURL
+	} else {
+		sendUrl = smsURL
+	}
+	for _, attachment := range msg.Attachments() {
+		_, mediaURL := handlers.SplitAttachment(attachment)
+		payload.MediaURLs = append(payload.MediaURLs, mediaURL)
+	}
+
+	jsonBody, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, sendUrl, bytes.NewReader(jsonBody))
+
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	var bearer = "AccessKey " + authToken
+	req.Header.Set("Authorization", bearer)
+
+	resp, respBody, err := handlers.RequestHTTP(req, clog)
+	if err != nil || resp.StatusCode/100 != 2 {
+		return status, nil
+	}
+	status.SetStatus(courier.MsgWired)
+	sendStatus := &ReceivedMessage{}
```

Comment:
A little preference is to put the types before the methods they are used 

Reusing `receiveMessage` seems confusing to me.

And probably it is best to call the types `MTPayload`, `MTResponse` and `MOPayload`
Even if MTPayload looks similar to MOPayload, we just add the fields we'll use one

Reply:
Actually here we could have `externalID, err := jsonparser.GetString(respBody, "id")` with using any type at all since we are just interested to get the ID only